plugins/plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/AddToIndexAction.java

```diff
@@ -13,38 +13,128 @@
                             AppContext appContext,
-                            GitLocalizationConstant constant) {
+                            GitLocalizationConstant constant,
+                            GitOutputConsoleFactory gitOutputConsoleFactory,
+                            ProcessesPanelPresenter consolesPanelPresenter,
+                            GitServiceClient service,
+                            NotificationManager notificationManager) {
         super(constant.addToIndexTitle(), constant.addToIndexTitle(), FontAwesome.PLUS, appContext);
         this.presenter = presenter;
+        this.constant = constant;
+        this.gitOutputConsoleFactory = gitOutputConsoleFactory;
+        this.consolesPanelPresenter = consolesPanelPresenter;
+        this.service = service;
+        this.notificationManager = notificationManager;
     }
 
     /** {@inheritDoc} */
     @Override
     public void actionPerformed(ActionEvent e) {
-        final Project project = appContext.getRootProject();
+        final Resource[] resources = appContext.getResources();
+        final DevMachine devMachine = appContext.getDevMachine();
+        final GitOutputConsole console = gitOutputConsoleFactory.create(ADD_TO_INDEX_COMMAND_NAME);
+        consolesPanelPresenter.addCommandOutput(devMachine.getId(), console);
+        service.getStatus(devMachine, appContext.getRootProject().getLocation())
+               .then(new Operation<Status>() {
+                   @Override
+                   public void apply(Status status) throws OperationException {
+                       if (containsInSelected(status.getUntracked())) {
+                           presenter.showDialog();
+                       } else if (containsInSelected(status.getModified())) {
+                           addToIndex(console);
+                       } else {
+                           String message = resources.length > 1 ? constant.nothingAddToIndexMultiSelect() : constant.nothingAddToIndex();
+                           console.print(message);
+                           notificationManager.notify(message);
+                       }
+                   }
+               })
+               .catchError(new Operation<PromiseError>() {
+                   @Override
+                   public void apply(PromiseError error) throws OperationException {
+                       console.printError(constant.statusFailed());
+                       notificationManager.notify(constant.statusFailed(), FAIL, FLOAT_MODE);
+                   }
+               });
+    }
 
-        checkState(project != null, "Null project occurred");
+    private void addToIndex(final GitOutputConsole console) {
+        Resource[] resources = appContext.getResources();
+        Path[] paths = new Path[resources.length];
+        for (int i = 0; i < resources.length; i++) {
+            Path path = resources[i].getLocation().removeFirstSegments(appContext.getRootProject().getLocation().segmentCount());
```

Comment:
`appContext.getRootProject().getLocation().segmentCount()` will be always equals to `1`, so you can safely remove the first segment without this checking.